lib/_stream_transform.js

```diff
@@ -217,3 +230,110 @@ function done(stream, er, data) {
+      if (source[kSourceIteratorResolve] === null) {
+        source[kSourceIteratorChunk] = chunk;
+        source[kSourceIteratorPull] = cb;
+        return;
+      }
+      source[kSourceIteratorResolve]({ value: chunk, done: false });
+      source[kSourceIteratorResolve] = null;
+      cb(null);
+    }
+  });
+  this.encoding = this[kSourceIteratorStream]._transformState.writeencoding;
+  this[kSourceIteratorGrabResolve] = (resolve) => {
+    this[kSourceIteratorResolve] = resolve;
+  };
+  this[kSourceIteratorPump]();
+}
+
+SourceIterator.prototype[Symbol.asyncIterator] = function() {
+  return this;
+};
+
+Object.setPrototypeOf(SourceIterator.prototype,
+                      AsyncIteratorPrototype);
+
+SourceIterator.prototype.next = function next() {
+  if (this[kSourceIteratorPull] === null || this[kSourceIteratorChunk] === null)
+    return new Promise(this[kSourceIteratorGrabResolve]);
+
+  this[kSourceIteratorPull](null);
+  const result = Promise.resolve({
+    value: this[kSourceIteratorChunk],
+    done: false });
+  this[kSourceIteratorChunk] = null;
+  this[kSourceIteratorPull] = null;
+  return result;
+};
+
+SourceIterator.prototype[kSourceIteratorPump] = async function pump() {
+  const stream = this[kSourceIteratorStream];
+  try {
+    const asyncGeneratorFn = this[kSourceIteratorGenerator];
+    const transformIter = asyncGeneratorFn(this);
+    stream.removeListener('prefinish', prefinish);
+    stream.on('prefinish', () => {
+      if (this[kSourceIteratorResolve] !== null) {
+        this[kSourceIteratorResolve]({ value: undefined, done: true });
+      }
+    });
+    let next = await transformIter.next();
+    do {
+      if (next.done) {
+        if (next.value !== undefined) stream.push(next.value);
+
+        // In the event of an early return we explicitly
+        // discard any buffered state
+        if (stream._writableState.length > 0) {
+          const { length } = stream._writableState;
+          const { transforming } = stream._transformState;
+          stream._writableState.length = 0;
+          stream._transformState.transforming = false;
+          prefinish.call(stream);
+          stream._writableState.length = length;
+          stream._transformState.transforming = transforming;
+        } else {
+          prefinish.call(stream);
+        }
+        break;
+      }
+      stream.push(next.value);
+    } while (next = await transformIter.next());
+  } catch (err) {
+    stream.destroy(err);
+  } finally {
+    this[kSourceIteratorPull] = null;
+    this[kSourceIteratorChunk] = null;
+    this[kSourceIteratorResolve] = null;
+    this[kSourceIteratorGenerator] = null;
+    this[kSourceIteratorStream] = null;
+  }
+};
+
+
+Transform.by = function by(asyncGeneratorFn, opts) {
+  if (!(asyncGeneratorFn instanceof AsyncGeneratorFunction))
+    throw new ERR_INVALID_ARG_TYPE(
+      'asyncGeneratorFn',
+      ['AsyncGeneratorFunction'],
+      asyncGeneratorFn);
```

Comment:
style nit: braces around multi-line `if`s

practical question: What if `asyncGeneratorFn instanceof AsyncGeneratorFunction` is false because of transpilation?

Reply:
(Or, if transpilation doesn’t apply, what if `asyncGeneratorFn` is from a different vm Context?)

Reply:
good point, how is this scenario typically handled according to node core convention?

Reply:
I’m pretty sure there’s no reliable way to detect what type of object a function will return, so I’d just leave the typecheck away here – if it’s not an async generator, it will throw an error later anyway

Reply:
hmm.. this is a larger issue though. I’m planning to allow pipeline to accept async function generators directly (talk to @mcollina for why this makes sense) - and in that case there’s no getting around the need to detect an async generator function. Is there any anticipated use cases where an async generator function would be leased from a vm context?

Reply:
Any function an async iterable. Iterator symbols exist for a reason. You should use `Symbol.asyncIterator` and expect people to call generators if they want to pass them in.

`Transform.by(iterator)` or `Transform.by(fnThatReturnsIterator())`

Reply:
@devsnek unfortunately that's not possible because we need to pass in the stream as an input parameter of the function:

```js
async function * mapper(source) { // <- note the source.
  for await (const chunk of source) {
    yield chunk.toUpperCase();
  }
}
```

This means we cannot support `Transform.by(iterator)` or `Transform.by(fnThatReturnsIterator())`

We could change the validation so that:

1. we check if it's a function.
2. we validate that the return value of that function has a `Symbol.asyncIterator` property, and call that function, failing if this does not exists.

Reply:
We could pass the source as the first yield result... 
but that would be a strange api for people. 

```
async function * mapper() {
  const source = yield;
  for await (const chunk of source) {
    yield chunk.toUpperCase();
  }
}
```

Reply:
resolved in latest commit